src/compiler/moduleSpecifiers.ts

```diff
@@ -832,71 +847,133 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode = MatchingMode.Exact): { moduleFileToTry: string; } | undefined {
+function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, host: ModuleSpecifierResolutionHost, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode, isImports: boolean): { moduleFileToTry: string; } | undefined {
     if (typeof exports === "string") {
+        const ignoreCase = !hostUsesCaseSensitiveFileNames(host);
+        const getCommonSourceDirectory = () => host.getCommonSourceDirectory();
+        const outputFile = isImports && getOutputJSFileNameWorker(targetFilePath, options, ignoreCase, getCommonSourceDirectory);

```

Comment:
One thing with these helpers is that they are called when you know its not `--outFile` .. I have really lost track of when this is allowed and when its not with all the module resolution options so not sure if that needs to be handled or not as well.

Reply:
`outFile` isn’t allowed in any module resolution mode that supports package.json `"imports"`, so it’s ok here.